File: flink-java-code/src/main/java/com/atguigu/day2/CoFlatMapFunctionExample.java
package com.atguigu.day2;

import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.co.CoFlatMapFunction;
import org.apache.flink.util.Collector;

public class CoFlatMapFunctionExample {
    public static void main(String[] args) throws Exception {
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStream<Tuple2<Integer, String>> stream1 = env.fromElements(
                new Tuple2<>(1, "aaaaa"),
                new Tuple2<>(2, "bbbbb")
        );

        DataStream<Tuple2<Integer, String>> stream2 = env.fromElements(
                new Tuple2<>(1, "ccccc"),
                new Tuple2<>(2, "ddddd")
        );

        stream1
                .keyBy(r -> r.f0)
                .connect(stream2.keyBy(r -> r.f0))
                .flatMap(new CoFlatMapFunction<Tuple2<Integer, String>, Tuple2<Integer, String>, String>() {
                    @Override
                    public void flatMap1(Tuple2<Integer, String> value, Collector<String> out) throws Exception {
                        out.collect(value.f1 + " 来自第一条流的元素发送两次");
                        out.collect(value.f1 + " 来自第一条流的元素发送两次");
                    }

                    @Override
                    public void flatMap2(Tuple2<Integer, String> value, Collector<String> out) throws Exception {
                        out.collect(value.f1 + " 来自第二条流的元素发送一次");
                    }
                })
                .print();

        env.execute();
    }
}

File: flink-scala-code/src/main/scala/com/atguigu/day2/CoFlatMapFunctionExample.scala
package com.atguigu.day2

import org.apache.flink.streaming.api.functions.co.CoFlatMapFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

object CoFlatMapFunctionExample {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    val stream1: KeyedStream[(Int, String), Int] = env
      .fromElements(
        (1, "aaaaa"),
        (2, "bbbbb")
      )
      .keyBy(r => r._1)

    val stream2: KeyedStream[(Int, String), Int] = env
      .fromElements(
        (1, "ccccc"),
        (2, "ddddd")
      )
      .keyBy(r => r._1)

    val connected: ConnectedStreams[(Int, String), (Int, String)] = stream1.connect(stream2)

    connected
      .flatMap(new MyFlatMapFunction)
      .print()

    env.execute()
  }

  class MyFlatMapFunction extends CoFlatMapFunction[(Int, String), (Int, String), String] {

    private var v: String = _

    override def flatMap1(value: (Int, String), out: Collector[String]): Unit = {
      out.collect(value._2 + " 来自第一条流的元素发送两次")
      out.collect(value._2 + " 来自第一条流的元素发送两次")
      if (v != null) {
        out.collect(v + value._2)
      } else {
        v = value._2
      }
    }

    override def flatMap2(value: (Int, String), out: Collector[String]): Unit = {
      out.collect(value._2 + " 来自第二条流的元素发送一次")
      if (v != null) {
        out.collect(v + value._2)
      } else {
        v = value._2
      }
    }
  }
}